odxtools/cli/compare.py

```diff
@@ -33,73 +36,82 @@
     deleted_services: List[DiagService]  # List of deleted services
-    renamed_service: List[Tuple[str, DiagService]]  # List of (old_name, DiagService)
-    changed_name_of_service: List[Tuple[str, DiagService]]  # Similar to renamed_service
-    changed_parameters_of_service: List[Tuple[str, DiagService, List[Union[str, int, float]]]]
+    renamed_service: list  # type: ignore[type-arg]
+    changed_name_of_service: List[Tuple[str, DiagService]]
+    changed_parameters_of_service: list[Any, Any, DiagService]  # type: ignore[type-arg]
 
 
 @dataclass
 class SpecsChangesVariants:
-    new_variants: List[str]
-    deleted_variants: List[str]
-    service_changes: ServiceSpecs
+    new_diagnostic_layers: List[str]
+    deleted_diagnostic_layers: List[DiagLayer]
+    service_changes: Dict[str, ServiceSpecs]
 
 
 class Display:
-    param_detailed: bool = False
-    obj_detailed: bool = False
+    param_detailed: bool
+    obj_detailed: bool
 
     def __init__(self) -> None:
         pass
 
-    def print_dl_changes(self, service_spec: ServiceSpecs) -> None:
-
-        if (service_spec["new_services"] or service_spec["deleted_services"] or
-                service_spec["changed_name_of_service"] or
-                service_spec["changed_parameters_of_service"]):
+    def print_dl_changes(self, service_specs: ServiceSpecs) -> None:
+        if (
+            service_specs.new_services
+            or service_specs.deleted_services
+            or service_specs.changed_name_of_service
+            or service_specs.changed_parameters_of_service
+        ):
+            assert isinstance(service_specs.diag_layer, str)
             rich_print()
             rich_print(
-                f"Changed diagnostic services for diagnostic layer '{service_spec['diag_layer']}' ({service_spec['diag_layer_type']}):"
+                f"Changed diagnostic services for diagnostic layer '{service_specs.diag_layer}' ({service_specs.diag_layer_type}):"
             )
-        if service_spec["new_services"]:
+        if service_specs.new_services:
+            assert isinstance(service_specs.new_services, List)
             rich_print()
             rich_print(" [blue]New services[/blue]")
-            rich_print(extract_service_tabulation_data(service_spec.new_services))
+            rich_print(extract_service_tabulation_data(service_specs.new_services))
 
-        if service_spec["deleted_services"]:
+        if service_specs.deleted_services:
+            assert isinstance(service_specs.deleted_services, List)
             rich_print()
             rich_print(" [blue]Deleted services[/blue]")
-            rich_print(extract_service_tabulation_data(service_spec["deleted_services"]))
-        if service_spec["changed_name_of_service"][0]:
+            rich_print(extract_service_tabulation_data(service_specs.deleted_services))
+        if service_specs.changed_name_of_service:
+            # assert isinstance(service_specs.changed_name_of_service[0], List)
```

Comment:
only leave such "ruins" in the code base in exceptional circumstances (IMO this is not one of these)